### PR TITLE
Allow permission requests within the same eTLD+1

### PR DESCRIPTION
**Pull request by bbondy**
Resolves https://github.com/brave/brave-browser/issues/23142

## Submitter Checklist:

- [x] I confirm that no security/privacy review [is needed](https://github.com/brave/brave-browser/wiki/Security-reviews), or that I have [requested](https://github.com/brave/security/issues/new/choose) one
  - Discussed it on Slack with Yan and Anthony here https://bravesoftware.slack.com/archives/G5SLA7LHW/p1653926667068809
- [x] There is a [ticket](https://github.com/brave/brave-browser/issues) for my issue
- [x] Used Github [auto-closing keywords](https://docs.github.com/en/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue) in the PR description above
- [x] Wrote a good [PR/commit description](https://google.github.io/eng-practices/review/developer/cl-descriptions.html)
- [x] Squashed any review feedback or "fixup" commits before merge, so that history is a record of what happened in the repo, not your PR
- [x] Added appropriate labels (`QA/Yes` or `QA/No`; `release-notes/include` or `release-notes/exclude`; `OS/...`) to the associated issue
- [x] Checked the PR locally: `npm run test -- brave_browser_tests`, `npm run test -- brave_unit_tests`, `npm run lint`, `npm run gn_check`, `npm run tslint`
- [x] Ran `git rebase master` (if needed)

## Reviewer Checklist:

- [ ] A security review [is not needed](https://github.com/brave/brave-browser/wiki/Security-reviews), or a link to one is included in the PR description
- [ ] New files have MPL-2.0 license header
- [ ] Adequate test coverage exists to prevent regressions
- [ ] Major classes, functions and non-trivial code blocks are well-commented
- [ ] Changes in component dependencies are properly reflected in `gn`
- [ ] Code follows the [style guide](https://chromium.googlesource.com/chromium/src/+/HEAD/styleguide/c++/c++.md)
- [ ] Test plan is specified in PR before merging

## After-merge Checklist:

- [ ] The associated issue milestone is set to the smallest version that the
  changes has landed on
- [ ] All relevant documentation has been updated, for instance:
  - [ ] https://github.com/brave/brave-browser/wiki/Deviations-from-Chromium-(features-we-disable-or-remove)
  - [ ] https://github.com/brave/brave-browser/wiki/Proxy-redirected-URLs
  - [ ] https://github.com/brave/brave-browser/wiki/Fingerprinting-Protections
  - [ ] https://github.com/brave/brave-browser/wiki/Brave%E2%80%99s-Use-of-Referral-Codes
  - [ ] https://github.com/brave/brave-browser/wiki/Custom-Headers
  - [ ] https://github.com/brave/brave-browser/wiki/Web-Compatibility-Exceptions-in-Brave
  - [ ] https://github.com/brave/brave-browser/wiki/QA-Guide
  - [ ] https://github.com/brave/brave-browser/wiki/P3A

### CODE REVIEW

File: components/permissions/contexts/brave_wallet_permission_context.cc [hunk 1]
```diff
@@ -163,9 +164,12 @@ void BraveWalletPermissionContext::RequestPermissions(
   }
 
   auto* web_contents = content::WebContents::FromRenderFrameHost(rfh);
-  // Fail the request came from 3p origin.
-  if (web_contents->GetMainFrame()->GetLastCommittedOrigin() !=
-      rfh->GetLastCommittedOrigin()) {
+  // Fail the request came from a different eTLD+1 origin.
+  if (!rfh->GetParent() &&
```
Comment (darkdh) [inline components/permissions/contexts/brave_wallet_permission_context.cc:168]: This should be `rfh->GetParent()` if we are looking for iframe.

File: components/permissions/contexts/brave_wallet_permission_context.cc [hunk 2]
```diff
@@ -211,9 +215,11 @@ void BraveWalletPermissionContext::GetAllowedAccounts(
   }
 
   auto* web_contents = content::WebContents::FromRenderFrameHost(rfh);
-  // Fail the request came from 3p origin.
-  if (web_contents->GetMainFrame()->GetLastCommittedOrigin() !=
-      rfh->GetLastCommittedOrigin()) {
+  if (rfh->GetParent() &&
```
Comment (darkdh) [inline components/permissions/contexts/brave_wallet_permission_context.cc:218]: `// Fail the request came from a different eTLD+1 origin.` is missing for this function

**Review by darkdh — APPROVED** [review]
lgtm, only one nit left.

File: browser/permissions/permission_manager_browsertest.cc [hunk 3]
```diff
@@ -341,6 +341,42 @@ IN_PROC_BROWSER_TEST_F(PermissionManagerBrowserTest,
...
+  for (size_t i = 0; i < sizeof(cases) / sizeof(cases[0]); ++i) {
+    // Will return empty responses without prompt.
+    BraveWalletPermissionContext::RequestPermissions(
+        cases[i].type, iframe_rfh, cases[i].addresses,
+        base::BindOnce([](const std::vector<ContentSetting>& responses) {
+          EXPECT_FALSE(responses.empty());
+        }));
+
+    content::RunAllTasksUntilIdle();
+    EXPECT_FALSE(IsPendingGroupedRequestsEmpty(cases[i].type)) << "case: " << i;
```
Comment (goodov) [inline browser/permissions/permission_manager_browsertest.cc:372]: can also write it like this, seems okay to distinguish cases by the wallet address:
```
for (const auto& case : cases) {
  SCOPED_TRACE(testing::PrintToString(case.addresses));
  ...
```

File: components/permissions/contexts/brave_wallet_permission_context.cc (hunk 1 above)
Comment (bbondy) [reply components/permissions/contexts/brave_wallet_permission_context.cc]: Fixed in the latest

File: components/permissions/contexts/brave_wallet_permission_context.cc (hunk 2 above)
Comment (bbondy) [reply components/permissions/contexts/brave_wallet_permission_context.cc:228]: Fixed in 9d0ff1f736f79bae8aed8eaf90607b24469e8d6b

File: browser/permissions/permission_manager_browsertest.cc [hunk 4]
```diff
@@ -341,6 +341,42 @@ IN_PROC_BROWSER_TEST_F(PermissionManagerBrowserTest,
...
+
+  for (size_t i = 0; i < sizeof(cases) / sizeof(cases[0]); ++i) {
+    // Will return empty responses without prompt.
+    BraveWalletPermissionContext::RequestPermissions(
+        cases[i].type, iframe_rfh, cases[i].addresses,
+        base::BindOnce([](const std::vector<ContentSetting>& responses) {
+          EXPECT_FALSE(responses.empty());
+        }));
+
+    content::RunAllTasksUntilIdle();
```
[the remark this answers is not in the document]
Comment (bbondy) [reply browser/permissions/permission_manager_browsertest.cc:401]: Fixed in 9d0ff1f736f79bae8aed8eaf90607b24469e8d6b

File: browser/permissions/permission_manager_browsertest.cc (hunk 3 above)
Comment (bbondy) [reply browser/permissions/permission_manager_browsertest.cc]: I didn't know about that, thanks! 
Fixed in 9d0ff1f736f79bae8aed8eaf90607b24469e8d6b

**Comment by bbondy** [comment]
FYI - I removed some debug prints and updated comments to reference the latest hash.

**Review by darkdh — DISMISSED** [review]
++

**Review by goodov — DISMISSED** [review]

File: chromium_src/chrome/browser/permissions/chrome_permissions_client.cc [hunk 5]
```diff
@@ -39,7 +40,9 @@ bool ChromePermissionsClient::BraveCanBypassEmbeddingOriginCheck(
           permissions::ContentSettingsTypeToRequestType(type),
           url::Origin::Create(requesting_origin), &original_requesting_origin,
           nullptr) &&
-      original_requesting_origin == url::Origin::Create(embedding_origin)) {
+      net::registry_controlled_domains::SameDomainOrHost(
+          original_requesting_origin, url::Origin::Create(embedding_origin),
+          net::registry_controlled_domains::INCLUDE_PRIVATE_REGISTRIES)) {
```
Comment (diracdeltas) [inline chromium_src/chrome/browser/permissions/chrome_permissions_client.cc:45]: just wanted to double check - `embedding_origin` is the top level (main frame) origin as opposed to the parent origin? ex: if a.com embeds b.com which embeds c.com and the requesting origin is c.com, embedding origin is a.com not b.com.

**Comment by bbondy** [comment]
Closing in favor of https://github.com/brave/brave-core/pull/13783

**Pull request closed without merging**